### ISSUE TITLE
The Rubinius process is aborting with signal: SIGSEGV

### ISSUE THREAD

**Issue by binaryplease** [rubinius/rubinius#3322]
Hi, I ran into this error:

http://pastie.org/9959209
https://gist.github.com/binaryplease/c398d495f84904dc11ab

The project is pretty big and I can't really figure out what the problem is. Using the default ruby implementation "ruby-head" doesnt throw this error.
Also the code runs a few minutes and stops then, so debugging is a bit problematic.
How do I fix this? How do I find out what causes this error?

Version:
rubinius 2.5.2 (2.1.0 7a5b05b1 2015-01-30 3.5.1 JI) [x86_64-linux-gnu]
OS:
Linux binaryplease-laptop 3.18.6-1-ARCH #1 SMP PREEMPT Sat Feb 7 08:44:05 CET 2015 x86_64 GNU/Linux

**Comment by YorickPeterse** [comment]
Are you using mysql2 in this project? Also, if you have GDB installed can you share a Gist of the GDB backtraces (see http://rubini.us/doc/en/how-to/obtaining-gdb-backtraces/ on how to do this).

**Comment by binaryplease** [comment]
Hi im using the mysql gem. not mysql2 though.

I followed the gdb instructions you gave. Here is the result:

https://gist.github.com/binaryplease/594c441e9bf1bbefddeb

**Comment by YorickPeterse** [comment]
@binaryplease Is it possible for you to try out mysql2 (https://github.com/brianmario/mysql2)? It's a more modern MySQL driver and they spent quite some time recently getting it up and running on Rubinius. 

Sadly the logs don't really provide much information due to the lack of debugging symbols. If you're up for it you can build Rubinius from source using `./configure --debug-build` (and any other options you might need) and try that. It will slow things down but you should at least get better stack traces

**Comment by binaryplease** [comment]
Ok, I'will try mysql2 then. It will take me a few days unill I get it done, then I will report back. Also, does sqeuel (http://sequel.jeremyevans.net/) work with rubinius?

**Comment by YorickPeterse** [comment]
The Sequel Gem works fine with Rubinius yes.

**Comment by binaryplease** [comment]
Ok ill try that one then. Just have to find and install a good mysql driver for it.

**Comment by brixen** [comment]
@binaryplease from your [backtrace](https://gist.github.com/binaryplease/c398d495f84904dc11ab#file-gistfile1-rb-L25), it appears that you could be running into an issue with thread-safety of `getaddrinf`, `gethostbyname`: https://encrypted.google.com/search?hl=en&q=gethostbyname%20getaddrinfo%20thread%20safe

**Comment by YorickPeterse** [comment]
@binaryplease You can use mysql2 with Sequel, but you can also use the former _without_ Sequel.

**Comment by binaryplease** [comment]
@brixen I don't use that functions explicitly. The might be called by nokogiri gem or by mechanize, but I would have assumed that they are thread-safe. Also the problem doesn't occur with standard mri and takes about 10 minutes to occur with rubinius.

**Comment by YorickPeterse** [comment]
@binaryplease Nokogiri doesn't do any domain resolution, so that is unlikely. Mechanize doesn't have a C extension so it's unlikely to be the culprit. Looking at the backtrace it's much more likely that the mysql Gem uses this to, probably, look up a server address.

> Also the problem doesn't occur with standard mri and takes about 10 minutes to occur with rubinius.

MRI has the GIL which prevents multiple threads from running in parallel (with the exception of some IO related things). It not occurring on MRI is most likely due to said GIL.

**Comment by committance** [comment]
Hi,
I tested the mysql2 gem without sequel. The error persists no change whatsoever.
The stacktrace is the same.

So how do I continue?

**Comment by binaryplease** [comment]
Hi,
I tested the mysql2 gem without sequel. The error persists no change whatsoever.
The stacktrace is the same.

So how do I continue?

**Comment by YorickPeterse** [comment]
Do you have some sort of sample script that we can use to reproduce the problem?

**Comment by binaryplease** [comment]
@YorickPeterse Thats the problem. The project is big and It runs a few minutes without problems before crashing.
I cant figure out where it crashes. If you see in the  backtrace which file/line is problemtatic I can look into it.

**Comment by brixen** [comment]
@binaryplease MRI has a global lock. Your Ruby threads are never running in parallel. So whether it works on MRI tells us nothing of value.

What I linked you to is likely the issue. Where that is getting called, I'm not certain without being able to run your code. Rubinius logs the backtrace of all the Ruby threads if it's able to do so. Your output shows this https://gist.github.com/binaryplease/c398d495f84904dc11ab. You'll need to go through those and see what code is being called and what that code calls.

**Comment by binaryplease** [comment]
@bixen I just made a test uninstalling any mysql gems and replacing my mysqlbackend methods with an simple array filled with test data. The error occured anyway. So th mysql gem can't be the issue.

**Comment by brixen** [comment]
@binaryplease given that you're running a crawler using Mechanize using Net::HTTP using DNS somewhere, I wouldn't be surprised that it's not mysql gem, nor that you have a potential threading issue.

The confusing thing is that most of what I read on google says that `getaddrinfo` is supposed to be thread-safe, but this backtrace would appear to contradict that https://gist.github.com/binaryplease/c398d495f84904dc11ab#file-gistfile1-rb-L18-L25.

I'm also curious why this frame says `ruby()` https://gist.github.com/binaryplease/c398d495f84904dc11ab#file-gistfile1-rb-L18 when the segv handler shows `rbx` as the program name.

Without being able to run your code and investigate this in gdb/lldb, there's not much I can do to help. You'll need to attach to the process with gdb/lldb and wait for the segv to occur and look at where all the threads are. Until you do this, you're just speculating.

**Comment by binaryplease** [comment]
@brixen I understand that its complicated without having the actual code, but I ran it with gdb, waiteed for the segv and posted the output as a gist above. I just dont really know how to use this backtrace properly. 

You mentioned that DNS might be a problem. Would it then be a solution to translate the hosts to IP adresses first and use mechanize with the IPs? That should gt rig of DNS if I understand correctly.

**Comment by brixen** [comment]
@binaryplease I looked more closely through the Ruby backtrace output:

Thread 3094: [Socket::Foreign.getaddrinfo in /home/binaryplease/.rvm/rubies/rbx-2.5.2/gems/gems/rubysl-socket-2.0.1/lib/rubysl/socket.rb:393 (+226 jit)](https://gist.github.com/binaryplease/c398d495f84904dc11ab#file-gistfile1-rb-L242)
Thread 3114: [Socket::Foreign.getaddrinfo in /home/binaryplease/.rvm/rubies/rbx-2.5.2/gems/gems/rubysl-socket-2.0.1/lib/rubysl/socket.rb:393 (+226 jit)](https://gist.github.com/binaryplease/c398d495f84904dc11ab#file-gistfile1-rb-L486)

The Socket::Foreign.getaddrinfo method is [creating a new MemoryPointer](https://github.com/rubysl/rubysl-socket/blob/2.0/lib/rubysl/socket.rb#L393) for the libc call, so I don't think it's an issue with Socket.

I was also going to suggest resolving the IP either before hand or as a service on a single thread that the other threads use.

**Comment by brixen** [comment]
@binaryplease it is definitely frustrating that docs appear to say `getaddrinfo` is thread safe but your backtrace appears to contradict that quite clearly.

**Comment by binaryplease** [comment]
Ok,
I changed my the implementation in my project to use only one mechanize agent per thread and it ran a lot longer before the error occured.

I can't use only one mechanize agent in the whole crawler because that would drop the performace too much as I need to retrieve more than one page at once if the server is responding slowly.

This is the stacktrace

https://gist.github.com/binaryplease/1599fe65cc7d328bac80

Maybe it helps narrowing down the problem.
I will now run a test with ip adresses instead of urls.

**Comment by binaryplease** [comment]
Sorry for double posting, but i found this:

https://coderrr.wordpress.com/2007/12/06/backgrounding-tasks-in-rails-with-threads-and-using-resolv-replace-to-make-them-faster/

I advises to use "require 'resolv-replace'" when working woth threads. I think this could be a solution, but I dont know how to use it, as requiring it gives me a load error and there is no gem by this name.

What do I need to install to test this? Where has it to be required, In the same class where I create my Mechanize object?

**Comment by brixen** [comment]
> I changed my the implementation in my project to use only one mechanize agent per thread and it ran a lot longer before the error occured.

@binaryplease if your "agent per thread" is calling this function on more than one thread at the same time, you have the same problem.

I'm not suggesting you use one mechanize agent per process. This is what I suggested:

> I was also going to suggest resolving the IP either before hand or as a service on a single thread that the other threads use.

In other words, you'd have N threads running mechanize but either 1) you resolve all hosts to IPs beforehand, or 2) you have a single thread that provides a service to resolve hosts to IPs based on requests from other threads, but only one thread in your process calls `getaddrinfo`.

**Comment by committance** [comment]
Hi,
I'm trying the solution you proposed. It seems to work, just not very good. I get no segv but the performance drops drastically. I also get problems redirecting and not finding some sites. I might have erros converting the url to an ip.

This is what I used:

    def ip(url)
        ip = '0.0.0.0'
        host = URI.parse(url).host
            
        begin
        @lock.synchronize do
            ip = IPSocket.getaddress(host)
        end
        rescue SocketError

            puts "Scoketerror #{url}|"

            return nil
        end
        url.gsub(host, ip)
    end


This is not a real solution for me though, since the performance is so low, I'm faster running default MRI. 

Is there a possibility to override the library mechanize uses to resolve the urls with something thread safe?

**Comment by binaryplease** [comment]
Hi,
I'm trying the solution you proposed. It seems to work, just not very good. I get no segv but the performance drops drastically. I also get problems redirecting and not finding some sites. I might have erros converting the url to an ip.

This is what I used:

    def ip(url)
        ip = '0.0.0.0'
        host = URI.parse(url).host
            
        begin
        @lock.synchronize do
            ip = IPSocket.getaddress(host)
        end
        rescue SocketError

            puts "Scoketerror #{url}|"

            return nil
        end
        url.gsub(host, ip)
    end


This is not a real solution for me though, since the performance is so low, I'm faster running default MRI. 

Is there a possibility to override the library mechanize uses to resolve the urls with something thread safe?

**Comment by YorickPeterse** [comment]
@binaryplease One option is to test if the original code works when using a global lock around any mechanize interactions, although performance would suffer from this as well. An example of such a lock for the AWS SDK: https://github.com/aws/aws-sdk-ruby/issues/455#issuecomment-48486235

**Comment by binaryplease** [comment]
Ok, Ill put a Mutex around the lines "agent.get(url)" and "agent.head(url)" that should ensure only one resolving process is running.

**Comment by binaryplease** [comment]
@YorickPeterse I created  muex in a global variable and used it with synchcronize on the relevant mechanize parts. The error didnt occur, but the performance is not usable. I am way slower than with MRI Ruby and no Mutex.

Is there a possibility to replace the dns resolver used in mechanize? I have been looking around but can't find a solution.

**Comment by YorickPeterse** [comment]
@binaryplease No idea to be honest. Interesting enough I've been running plenty of network related applications on Rubinius over time and I've never bumped into this particular problem. I doubt it's platform related since we're both running Arch Linux (at least by the looks of it).

If you could cook up some small script that reproduces this some way that would be of great help to us in finding out what exactly is going on.

**Comment by binaryplease** [comment]
https://books.google.de/books?id=E_KMa5Zab0MC&pg=PA507&lpg=PA507&dq=ruby+resolv::dns+thread+safe&source=bl&ots=eA6H4fuERa&sig=ETp4i95ns0LIAC8ZIs2kO3aKNaE&hl=de&sa=X&ei=eVjrVN_sB9DkatGUgsgL&ved=0CCsQ6AEwAQ#v=onepage&q=ruby%20resolv%3A%3Adns%20thread%20safe&f=false

Just found this. It is for ruby 1.9 but it says to include 'resolv' and 'resolv-replace' to make dns requests thread safe.

I can require 'resolv' but 'resolv-replace' gives me a load error in rubinius. ( Not in MRI though) Maybe this could be a solution.
Yes, Im using Arch. I'll try to cook up a script, but Its a bit problematic. If I get the error from it I'll let you know.

**Comment by YorickPeterse** [comment]
resolv-replace looks easy enough to copy into rubysl-resolv: https://github.com/ruby/ruby/blob/trunk/lib/resolv-replace.rb

I'll see if I can copy this into the Gem, maybe add some tests (if worth the pain) and release it somewhere later this night.

**Comment by binaryplease** [comment]
@YorickPeterse Cool. I don't really know how to do it myself, but If you let me know, I'll be happy to help testing and see if it solves the problem.

**Comment by binaryplease** [comment]
This might be relevant:
https://github.com/rubinius/rubinius/issues/2303

**Comment by YorickPeterse** [comment]
@binaryplease I updated and release a new version of rubysl-resolv which contains resolv-replace. Sadly for the time being you'll have to build Rubinius from source (until the next release) as pre-packaged rubysl Gems take priority over Gem installed ones.

**Comment by binaryplease** [comment]
@YorickPeterse Can you link me to some instructions for setting this up in Arch?

**Comment by YorickPeterse** [comment]
@binaryplease We can't really help with RVM, but building Rubinius from source itself is just a matter of running the following:

    git clone git@github.com:rubinius/rubinius.git
    cd rubinius
    bundle install
    ./configure --prefix=path/to/directory/to/install/rubinius/in
    rake install

If you have clang installed I recommend using the following configure command instead:

    ./configure --prefix=path/to/directory/to/install/rubinius/in --cc=clang --cxx=clang++

From the top of my head I think you can do something like `rvm mount path/to/directory/to/install/rubinius/in` but I'm not really sure as I don't use RVM myself.

**Comment by binaryplease** [comment]
Ok I installed as described.

But I have an error with ip_address_list 

    NoMethodError: undefined method `ip_address_list' on Socket (Class)


running this in irb is enough to reproduce:

    require 'socket'
    ip = Socket.ip_address_list.detect{|intf| intf.ipv4_private?}

**Comment by YorickPeterse** [comment]
Having already taken a look at this in the past I'll see if I can get said method implemented during the week.

**Comment by binaryplease** [comment]
@YorickPeterse Short update: I just tried it out with jruby, which also has native threads and has resolv-replace implemented. It solves the problem, without any other changes than require 'resolv-replace'

**Comment by YorickPeterse** [comment]
@binaryplease JRuby is an entirely different beast, it not occurring on JRuby doesn't instantly mean the problem is implementation specific.

**Comment by binaryplease** [comment]
@YorickPeterse Ok, ignore it then. I just thought It would show that resolv-replace solves the problem as the error also occures on jruby if I dont use resolv-replace.

**Comment by YorickPeterse** [comment]
The thing is that while their resolv/resolv-replace code might be the same Ruby code the underlying Socket internals might not be. That difference might ensure JRuby not to suffer from the same problem while Rubinius (and MRI if it wasn't for the GIL) will. As mentioned above I'll see if I can get Socket up to shape this week, but it will take a few days.

**Comment by YorickPeterse** [comment]
http://blog.gregburek.com/2015/02/22/dns-eglibc-and-resolv-replace-on-heroku/ This might be worth the read, giving more reasons to use resolv-replace (once I've finished implementing it).

**Comment by bruceadams** [comment]
I hit a similar issue, running the very simple set of threading tests that are part of https://github.com/bruceadams/pmap
Here is the crash I hit https://gist.github.com/bruceadams/e4ebe6be3c38575a0685

I was running the `pmap` tests repeatedly trying to reproduce a different issue. The test suite had run, and passed the tests, 4381 times before this crash. So, the tests are simple, but the crash is uncommon.

(If my crash is not similar enough to be treated as the same as this issue, I'm happy to open a new issue.)

**Comment by brixen** [comment]
Much of the internals of Rubinius have been completely or mostly rewritten in the past couple years. This includes the garbage collector, concurrency facilities, Fibers, much of the instruction set, and a migration away from "primitive" functions that implement Ruby features.

Since a number of segfaults or process hangs have occurred in these features over time, this issue may be fixed.

The focus for Rubinius in the near term is on the following capabilities:

1. Instruction set
2. Debugger
3. Profiler
4. Just-in-time compiler
5. Concurrency
6. Garbage collector

Contributions in the form of PRs for any of the areas of focus above are appreciated. Once these capabilities are more robust, it will be possible to more efficiently debug and fix any process crashes.

Other than these core capabilities, PRs to fix any specific issue are always welcome.
